### PR TITLE
Bias PPE's towards definition being printed

### PR DESCRIPTION
**Pull request by ChrisPenner**
Built on #3194 

It's long been an issue in unison where, when viewing a definition, some of the names might be unexpectedly from some other definition, and sometimes even the name of the thing you asked for gets replaced with a different name for that hash;

E.g. if I type `view trunk.List.map` it might print everything using the name `a1.List.map` instead, and even though those ARE valid names for the hashes, it's not what I asked for.

This fixes that by:

* When printing something the user requested, prefer names with the _longest common name-segment prefix_ with the name being printed.

E.g. if you're printing `base.trunk.List.foo`, which mentions `base.trunk.List.bar`, `lengthCommonPrefix(base.trunk.List.foo, base.trunk.List.bar) == 3` whereas `lengthCommonPrefix(base.trunk.List.foo, base.latest.List.bar) == 1`, so `base.trunk.List.bar` is preferred in the pretty printer, which is nice, because it's closest to what we're actually trying to print.

If two names have the same common-prefix length, we fall back on the current name-selection strategies.

## viewing `trunk.Seq` on `distributed` before:

image

## viewing `trunk.Seq` on `distributed` after:

image

## Implementation notes

When creating a PPE from a Names you may now pass a `bias` target, names close to that target will be prioritized.

## Interesting/controversial decisions

It's maybe a little messy to thread the bias argument through everywhere, but the results are good, so it's probably worth it 👍🏼

### CODE REVIEW

File: unison-src/transcripts/merges.output.md [hunk 1]
```diff
@@ -245,8 +245,8 @@ master.frobnicate n = n + 1
 
 .> view master.y
 
-  feature2.y : Text
-  feature2.y = "updated y"
+  master.y : Text
```
Comment (ChrisPenner) [inline unison-src/transcripts/merges.output.md:248]: 🎉

Comment (aryairani) [reply unison-src/transcripts/merges.output.md:248]: Love this

**Review by aryairani — APPROVED** [review]

File: unison-core/src/Unison/NamesWithHistory.hs [hunk 2]
```diff
@@ -250,32 +250,34 @@ termName length r NamesWithHistory {..} =
...
           where
             f fqn =
               let n' = Name.shortestUniqueSuffix fqn r rel
                   isHQ'd = R.manyDom fqn rel -- it is conflicted
-                  hq n = HQ'.take length (hq' n r)
+                  hq n = HQ'.take hashLen (hq' n r)
                   hqn = if isHQ'd then hq n' else HQ'.fromName n'
-               in (isHQ'd, Name.countSegments fqn, Name.countSegments n', Name.isAbsolute n', hqn)
+                  biasProximity n = negate . length . Name.commonPrefix n <$> mayBias
+               in (isHQ'd, biasProximity fqn, Name.countSegments fqn, Name.countSegments n', Name.isAbsolute n', hqn)
```
Comment (ChrisPenner) [inline unison-core/src/Unison/NamesWithHistory.hs]:
```suggestion
               in (biasProximity fqn, isHQ'd, Name.countSegments fqn, Name.countSegments n', Name.isAbsolute n', hqn)
```

**Comment by mitchellwrosen** [comment]
Hmm - oops, not type-checking - ping @ChrisPenner, I'll hold off on reviewing

**Comment by ChrisPenner** [comment]
Ack, yup looks like things changed on trunk and I'll need to rewrite some parts of this 😞

**Comment by ChrisPenner** [comment]
Yeah this doesn't play very nicely with the new 'current-namespace' bias that Paul added. I think we still want this feature, but I'll have to re-think how this fits with it. I have some other changes in mind for PPEs that I think will make all of this much simpler.

I'll put this on ice for the time being...

**Comment by pchiusano** [comment]
Ah, that's too bad. Should we move this off M4 then?

**Comment by ChrisPenner** [comment]
Closing in favour of https://github.com/unisonweb/unison/pull/3260 which does the same thing, but works properly with Paul's recent changes.

**Pull request closed without merging**